### PR TITLE
clear timeline events prior to starting a new action

### LINKED ISSUE #12982: FlutterDriver.traceAction should clear previous timeline events
Currently it returns old events, not just those that happen inside the action. That seems to lead to a false expectation that previous events will always be available, which is not the case.

### PR DESCRIPTION
**Pull request by yjbanov**
Fixes https://github.com/flutter/flutter/issues/12982

/cc @tvolkert @mehmetf

### CODE REVIEW

File: packages/flutter_driver/lib/src/driver.dart [hunk 1]
```diff
@@ -545,12 +549,38 @@ class FlutterDriver {
...
+  Future<Null> clearTimeline({ Duration timeout: _kShortTimeout }) async {
+    try {
+      await _peer
+          .sendRequest(_kClearVMTimelineMethod, <String, String>{})
+          .timeout(timeout);
+    } catch(error, stackTrace) {
+      throw new DriverError(
+          'Failed to clear event timeline due to remote error',
+          error,
+          stackTrace
```
Comment (gspencergoog) [inline packages/flutter_driver/lib/src/driver.dart]: trailing comma

**Review by gspencergoog — APPROVED** [review]

**Comment by gspencergoog** [comment]
lgtm

File: packages/flutter_driver/lib/src/driver.dart [hunk 2]
```diff
@@ -545,12 +549,38 @@ class FlutterDriver {
...
+  ///
+  /// If [retainPriorEvents] is true, retains events recorded prior to calling
+  /// [action]. Otherwise, prior events are cleared before calling [action]. By
+  /// default, prior events are cleared.
+  Future<Timeline> traceAction(
+    Future<dynamic> action(), {
+    List<TimelineStream> streams: _defaultStreams,
+    bool retainPriorEvents: false,
+  }) async {
+    if (!retainPriorEvents) {
```
Comment (tvolkert) [inline packages/flutter_driver/lib/src/driver.dart]: No curly braces for single-line `if` statements (have we relaxed than style requirement yet?)

File: packages/flutter_driver/lib/src/driver.dart [hunk 3]
```diff
@@ -545,12 +549,38 @@ class FlutterDriver {
...
+      await clearTimeline();
+    }
     await startTracing(streams: streams);
     await action();
     return stopTracingAndDownloadTimeline();
   }
 
+  /// Clears all timeline events recorded up until now.
+  Future<Null> clearTimeline({ Duration timeout: _kShortTimeout }) async {
```
Comment (tvolkert) [inline packages/flutter_driver/lib/src/driver.dart]: `@visibleForTesting`?

**Review by tvolkert — APPROVED** [review]
LGTM

File: packages/flutter_driver/lib/src/driver.dart (hunk 2 above)
Comment (yjbanov) [reply packages/flutter_driver/lib/src/driver.dart]: Yes, the style guide now says:

```dart
// ALSO GOOD:
if (notReady) {
  return;
}
```

File: packages/flutter_driver/lib/src/driver.dart (hunk 3 above)
Comment (yjbanov) [reply packages/flutter_driver/lib/src/driver.dart]: No, I think it's useful to be in the public API.

File: packages/flutter_driver/lib/src/driver.dart (hunk 1 above)
Comment (yjbanov) [reply packages/flutter_driver/lib/src/driver.dart]: Done (here and elsewhere)

**Pull request merged**